skrub/_datetime_encoder.py

```diff
@@ -18,78 +22,489 @@
+        Dictionary mapping parsable column indices to their datetime format.
+    """
+    indices = []
+    index_to_format = {}
+
+    for col_idx, X_col in enumerate(X_split):
+        X_col = X_col[pd.notnull(X_col)]
+
+        # convert pd.TimeStamp to np.datetime64
+        if all(isinstance(val, pd.Timestamp) for val in X_col):
+            X_col = X_col.astype("datetime64")
+
+        if _is_column_datetime_parsable(X_col):
+            indices.append(col_idx)
+
+            if np.issubdtype(X_col.dtype, np.datetime64):
+                # We don't need to specify a parsing format
+                # for columns that are already of type datetime64.
+                datetime_format = None
+            else:
+                datetime_format = _guess_datetime_format(X_col)
+
+            index_to_format[col_idx] = datetime_format
+
+    return indices, index_to_format
+
+
+def _is_column_datetime_parsable(X_col):
+    """Check whether a 1d array can be converted into a \
+    :class:`pandas.DatetimeIndex`.
+
+    Parameters
+    ----------
+    X_col : array-like of shape ``(n_samples,)``
+
+    Returns
+    -------
+    is_dt_parsable : bool
+    """
+    # Remove columns of int, float or bool casted as object.
+    # Pandas < 2.0.0 raise a deprecation warning instead of an error.
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", category=DeprecationWarning)
+        try:
+            if np.array_equal(X_col, X_col.astype(np.float64)):
+                return False
+        except ValueError:
+            pass
+
+    np_dtypes_candidates = [np.object_, np.str_, np.datetime64]
+    is_type_datetime_compatible = any(
+        np.issubdtype(X_col.dtype, np_dtype) for np_dtype in np_dtypes_candidates
+    )
+    if is_type_datetime_compatible:
+        try:
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", category=UserWarning)
+                # format=mixed parses entries individually,
+                # avoiding ValueError when both date and datetime formats
+                # are present.
+                # At this stage, the format itself doesn't matter.
+                _ = pd.to_datetime(X_col, format=MIXED_FORMAT)
+            return True
+        except (pd.errors.ParserError, ValueError):
+            pass
+    return False
+
+
+def _guess_datetime_format(X_col):
+    """Infer the format of a 1d array.
+
+    This functions uses Pandas ``guess_datetime_format`` routine for both
+    dayfirst and monthfirst case, and select either format when using one
+    give a unify format on the array.
+
+    When both dayfirst and monthfirst format are possible, we select
+    monthfirst by default.
+
+    You can overwrite this behaviour by setting a format of the caller function.
+    Setting a format always take precedence over infering it using
+    ``_guess_datetime_format``.
+
+    Parameters
+    ----------
+    X_col : ndarray of shape ``(n_samples,)``
```

Comment:
Upping Jerome's comment "If missing values are not allowed in X_col I think we should mention it in the docstring"